Sources/PackageLoading/ManifestLoader+Validation.swift

```diff
@@ -328,3 +321,29 @@ extension TargetDescription {
     fileprivate var isRemote: Bool { url != nil }
     fileprivate var isLocal: Bool { path != nil }
 }
+
+extension PackageDependency {
+    fileprivate var descriptionForValidation: String {
+        var description = self.nameForTargetDependencyResolutionOnly
+
+        if let locationsString = {
+            switch self {
+            case .fileSystem(let settings):
+                return settings.path.pathString
+            case .sourceControl(let settings):
+                switch settings.location {
+                case .local(let path):
+                    return path.pathString
+                case .remote(let url):
+                    return url.absoluteString
+                }
+            case .registry:
+                return .none
+            }
+        }() {
+            description += " (\(locationsString))"
```

Comment:
I think we need to workshop this a bit more

> valid dependencies are: 'foo (http://scm.com/org/foo)'

This looks extremely odd to me, how about something like:

> valid dependencies are: 'foo' (at 'http://scm.com/org/foo')

Reply:
maybe `at` for path, and `from` for SCM?